src/x509_extensions.rs

```diff
@@ -1,6 +1,514 @@
+    pub other: Vec<Oid<'a>>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct InhibitAnyPolicy {
+    /// The value indicates the number of additional non-self-issued
+    /// certificates that may appear in the path before anyPolicy
+    /// is no longer permitted.
+    pub skip_certs: u32,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PolicyMappings<'a> {
+    pub mappings: HashMap<Oid<'a>, Vec<Oid<'a>>>, 
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PolicyConstraints {
+    /// If present, number of additional certificates
+    /// that may appear in the path before an explicit policy is required for
+    /// the entire path.
+    pub require_explicit_policy: Option<u32>,
+    /// If present, number of additional certificates that may appear in the
+    /// path before policy mapping is no longer permitted.
+    pub inhibit_policy_mapping: Option<u32>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct SubjectAlternativeName<'a> {
+    pub general_names: Vec<GeneralName<'a>>,
+}
+
+#[derive(Debug, PartialEq)]
+/// Represents a GeneralName as defined in RFC5280. There
+/// is no support X.400 addresses and EDIPartyName.
+///
+/// String formats are not validated.
+pub enum GeneralName<'a> {
+    OtherName(Oid<'a>, &'a [u8]),
+    /// More or less an e-mail, the format is not checked.
+    RFC822Name(&'a str),
+    /// A hostname, format is not checked.
+    DNSName(&'a str),
+    // X400Address,
+    /// RFC5280 defines several string types, we always try to parse as utf-8
+    /// which is more or less a superset of the string types.
+    DirectoryName(crate::x509::X509Name<'a>),
+    // EDIPartyName { name_assigner: Option<&'a str>, party_name: &'a str },
+    /// An uniform resource identifier. The format is not checked.
+    URI(&'a str),
+    /// An ip address, provided as encoded.
+    IPAddress(&'a [u8]),
+    RegisteredID(Oid<'a>),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct NameConstraints<'a> {
+    pub permitted_subtrees: Option<Vec<GeneralSubtree<'a>>>,
+    pub excluded_subtrees: Option<Vec<GeneralSubtree<'a>>>,
+}
+
+#[derive(Debug, PartialEq)]
+/// Represents the structure used in the name contraints extensions.
+/// The fields minimum and maximum are not supported (openssl also has no support).
+pub struct GeneralSubtree<'a> {
+    pub base: GeneralName<'a>,
+    // minimum: u32,
+    // maximum: Option<u32>,
+}
+
+pub(crate) mod parser {
+    use nom::{alt, eof, exact, do_parse, opt, verify, call, take, many1, IResult, Err};
+    use der_parser::{*, oid::Oid};
+    use der_parser::ber::BerObject;
+    use der_parser::der::*;
+    use der_parser::error::BerError;
+
+    use crate::x509_extensions::*;
+
+    pub(crate) fn parse_extension_type<'a>(mut i: &'a [u8], oid: &Oid) -> IResult<&'a [u8], Option<ExtensionType<'a>>, BerError> {
+        let ext = if oid.asn1() == oid!(raw 2.5.29.15) {
+            let (ret, ku) = parse_keyusage(i)?;
+            i = ret;
+            Some(ExtensionType::KeyUsage(ku))
```

Comment:
Since `parse_keyusage` and similar are not exported, it would be easier to make them return an `ExtensionType`

This way, the `match` would be even more simplified as something like:
```rust
match oid.bytes() {
  OID_KEYUSAGE => parse_keyusage(i),
  // ...
}